L1Trigger/TrackFindingTracklet/src/PurgeDuplicate.cc

```diff
@@ -505,13 +534,165 @@ double PurgeDuplicate::getPhiRes(Tracklet* curTracklet, const Stub* curStub) {
+
+  // exact helix
+  if (st->isBarrel()) {
+    stub_r = settings_.rmean(stubLayer - 1);
+
+    double sin_val = (stub_r * stub_r + rho_minus_d0 * rho_minus_d0 - rho * rho) / (2 * stub_r * rho_minus_d0);
+    stub_phi = tracklet->phi0() - std::asin(sin_val);
+    stub_phi = stub_phi + iSector * settings_.dphisector() - 0.5 * settings_.dphisectorHG();
+    stub_phi = reco::reduceRange(stub_phi);
+
+    double beta = std::acos((rho * rho + rho_minus_d0 * rho_minus_d0 - stub_r * stub_r) / (2 * rho * rho_minus_d0));
+    stub_z = tracklet->z0() + tracklet->t() * std::abs(rho * beta);
+  } else {
+    stub_z = settings_.zmean(stubDisk - 1) * tracklet->disk() / abs(tracklet->disk());
+
+    double beta = (stub_z - tracklet->z0()) / (tracklet->t() * std::abs(rho));  // maybe rho should be abs value
+    double r_square = -2 * rho * rho_minus_d0 * std::cos(beta) + rho * rho + rho_minus_d0 * rho_minus_d0;
+    stub_r = sqrt(r_square);
+
+    double sin_val = (stub_r * stub_r + rho_minus_d0 * rho_minus_d0 - rho * rho) / (2 * stub_r * rho_minus_d0);
+    stub_phi = tracklet->phi0() - std::asin(sin_val);
+    stub_phi = stub_phi + iSector * settings_.dphisector() - 0.5 * settings_.dphisectorHG();
+    stub_phi = reco::reduceRange(stub_phi);
+  }
+
+  // TMP: for displaced tracking, exclude one of the 3 seeding stubs
+  // to be discussed
+  int seed = tracklet->seedIndex();
+  if ((seed == 8 && stubLayer == 4) || (seed == 9 && stubLayer == 5) || (seed == 10 && stubLayer == 3) ||
+      (seed == 11 && abs(stubDisk) == 1)) {
+    stub_phi = st->l1tstub()->phi();
+    stub_z = st->l1tstub()->z();
+    stub_r = st->l1tstub()->r();
+  }
+
+  std::vector invented_coords{stub_r, stub_z, stub_phi};
+  return invented_coords;
+}
+
+std::vector<const Stub*> PurgeDuplicate::seedStubCoordsFromTracklet(unsigned int iSector,
+                                                                    Tracklet* tracklet,
+                                                                    std::vector<const Stub*> originalStubsList) {
+  std::vector<const Stub*> newStubList;
+
+  for (unsigned int stubit = 0; stubit < originalStubsList.size(); stubit++) {
+    const Stub* thisStub = originalStubsList[stubit];
+
+    if (isSeedingStub(tracklet->seedIndex(), (findLayerDisk(thisStub)).first, (findLayerDisk(thisStub)).second)) {
+      // get a vector containing r, z, phi
+      std::vector<double> inv_r_z_phi;
+      if (!settings_.extended())
+        inv_r_z_phi = getInventedCoords(iSector, thisStub, tracklet);
+      else {
+        inv_r_z_phi = getInventedCoordsExtended(iSector, thisStub, tracklet);
+      }
+      double stub_x_invent = inv_r_z_phi[0] * std::cos(inv_r_z_phi[2]);
+      double stub_y_invent = inv_r_z_phi[0] * std::sin(inv_r_z_phi[2]);
+      double stub_z_invent = inv_r_z_phi[1];
+
+      Stub* invent_stub_ptr = new Stub(*thisStub);
+      const L1TStub* L1stub = thisStub->l1tstub();
```

Comment:
All variable names in CMSSW must start with a lower case letter.

Reply:
done